active/0000-service-workers.md

```diff
@@ -0,0 +1,48 @@
+- Start Date: 2018-03-13
+- RFC PR: (leave this empty)
+
+# Add a Service Worker with minimal asset caching to a default Ember CLI application
+## Summary
+This RFC introduces a Service Worker to the default app blueprint. It will only cache the most essential assets, such as `index.html`, and the vendor and app specific JS and CSS files.
+
+## Motivation
+A Service Worker allows you to cache assets and other resources in a  way that loading a page becomes reliable and fast, independent of what the network conditions are.
+
+Not all Ember applications are destined to become a Progressive Web App, but that doesn't mean that applications shouldn't leverage a simple Service Worker script to somewhat improve the reliability of loading it.
+
+## Detailed design
+### Registration
+A Service Worker script named `sw.js` should be compiled and placed at the root level of the build result. The Service Worker script should be registered with a registration script that is inlined in `index.html`. The registered Service Worker should have the `rootURL` that is configured in `config/environment.js` as scope.
+
+### Caching
+The default Service Worker configuration should only cache the following files:
+
+* `index.html`
+* `assets/vendor.js`
+* `assets/<app-name>.js`
+* `assets/vendor.css`
+* `assets/<app-name>.css`
+
+The cached `index.html` file should be served by the Service Worker whenever a `fetch` event happens that has the request mode of `navigation` and the `accept` header of the request has `text/html` as value.
+
+### Service Worker invalidation
+By default each build should include a random sequence of text in the compiled output of `sw.js`, this way the Service Worker gets invalidated after every build.
+
+### Kill switch
+The build process should be able to take a flag that acts as a kill switch in case the Service Worker needs to be unregistered in the deployed environment as the result of a bad deploy.
+
+When this flag is given, the registration script for the Service Worker should be altered to immediately unregister the current registered Service Worker.
+
+## How we teach this
+The Ember CLI documentation should be updated with a notice that a default Ember CLI application registers a Service Worker and explain the functionality of said Service Worker.
+
+It should also include a warning that a Service Worker must be served over HTTPS and served from the same origin to function correctly.
+
+It might also be worth to add a few pointers on how to set up developer tools to improve the development experience with an active Service Worker.
```

Comment:
We've been working with a service worker enabled app for a few months and have found most development tasks (particularly TDD) to be so painful that we all just run `SW_DISABLED=true ember serve`. While there are some workarounds; it shouldn't be underestimated how much complexity an extra cache layer can add to an active development process.  Consider turning this off by default in non-production environments.

Reply:
Instead of using `SW_DISABLED=true` most browsers have good dev tools to deal with this problem. For example in Chrome you can check `Update on reload` in the dev tools and it will will run the whole SW download/install/activate cycle before it loads the page.

Reply:
There are other considerations here -- for example, working with a variety of apps on `localhost:4200`, some of which are not PWAs. It's common for even very experienced developers, well versed in PWAs and the appropriate devtools best practices to regularly troll themselves.

Reply:
Yeah like the page is white because it loaded the SW for another app. Another reason I use etc/hosts to setup local test domains.